cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDynamicDirectedGraph.scala

```diff
@@ -0,0 +1,206 @@
+  // (See above note on outboundLists)
+  private val inboundLists = new ArrayBuffer[IntArrayList]
+
+  private var nodeCount_ = 0
+
+  def this(dataIterator: Iterator[NodeIdEdgesMaxId],
+            storedGraphDir: StoredGraphDir) {
+    this(storedGraphDir)
+    for (nodeData <- dataIterator) {
+      val id = nodeData.id
+      getOrCreateNode(id)
+      nodeData.edges map getOrCreateNode
+      storedGraphDir match {
+        case OnlyOut => outboundLists(id).addAll(IntArrayList.wrap(nodeData.edges))
+        case OnlyIn => inboundLists(id).addAll(IntArrayList.wrap(nodeData.edges))
+        case BothInOut => nodeData.edges map { addEdgeAllowingDuplicates(id, _) } // Duplicates shouldn't exist, but allow them for efficiency.
+      }
+    }
+  }
+
+  def this(iteratorSeq: Seq[() => Iterator[NodeIdEdgesMaxId]],
+           storedGraphDir: StoredGraphDir) {
+    this((iteratorSeq map { _()} ).iterator.flatten, storedGraphDir)
+  }
+
+  /* Returns an option which is non-empty if outbound list for id  is non-null. */
+  private def outboundListOption(id: Int): Option[IntArrayList] =
+    if (id < 0 || id >= outboundLists.size)
+      None
+  else
+      Option(outboundLists(id)) // If list is null, convert to None
+
+  private def inboundListOption(id: Int): Option[IntArrayList] =
+    if (id < 0 || id >= inboundLists.size)
+      None
+    else
+      Option(inboundLists(id)) // If list is null, convert to None
+
+  /** Wraps outboundList and inboundList in a node structure.  This is an inner class
+    * because mutations will affect other nodes of the graph if both inbound
+    * and outbound neighbor lists are stored.
+    *
+    * For efficiency, we don't store Nodes, but create them on the fly as needed.
+    */
+  class IntListNode(override val id: Int,
+                    val outboundList: Option[IntArrayList],
+                    val inboundList: Option[IntArrayList])
+      extends DynamicNode {
+    override def outboundNodes(): Seq[Int] = outboundList match {
+      case Some(list) => FastUtilUtils.intArrayListToSeq(list)
+      case None => Seq.empty[Int]
+    }
+
+    override def inboundNodes(): Seq[Int] = inboundList match {
+      case Some(list) => FastUtilUtils.intArrayListToSeq(list)
+      case None => Seq.empty[Int]
+    }
+
+    def addOutBoundNodes(nodeIds: Seq[Int]): Unit = {
+      //For future optimization, we could check if we are only storing outbound
+      // nodes and then do:
+      // outboundList.get.addAll(IntArrayList.wrap(nodeIds.toArray))
+      nodeIds map { addEdge(id, _) }
+    }
+
+    def addInBoundNodes(nodeIds: Seq[Int]): Unit = {
+      nodeIds map { addEdge(_, id) }
+    }
+
+    def removeOutBoundNode(nodeId: Int): Unit = removeEdge(id, nodeId)
+
+    def removeInBoundNode(nodeId: Int): Unit =  removeEdge(nodeId, id)
+  }
+
+  /**
+   * Determine if a node with the given id exists in this graph.
+   */
+  def nodeExists(id: Int): Boolean =
```

Comment:
I think outboundListOption already handles the case when id > outboundLists.length.  I just added
    graph.nodeExists(1000000) shouldEqual false
to one of the tests (where graph is a small graph), and it passed.